store/tikv/batch_coprocessor.go

```diff
@@ -0,0 +1,422 @@
+}
+
+type batchCopResponse struct {
+	pbResp *coprocessor.BatchResponse
+	detail *execdetails.ExecDetails
+
+	// batch Cop Response is yet to return startKey. So batchCop cannot retry partially.
+	startKey kv.Key
+	err      error
+	respSize int64
+	respTime time.Duration
+}
+
+// GetData implements the kv.ResultSubset GetData interface.
+func (rs *batchCopResponse) GetData() []byte {
+	return rs.pbResp.Data
+}
+
+// GetStartKey implements the kv.ResultSubset GetStartKey interface.
+func (rs *batchCopResponse) GetStartKey() kv.Key {
+	return rs.startKey
+}
+
+// GetExecDetails is unavailable currently, because TiFlash has not collected exec details for batch cop.
+// TODO: Will fix in near future.
+func (rs *batchCopResponse) GetExecDetails() *execdetails.ExecDetails {
+	return &execdetails.ExecDetails{}
+}
+
+// MemSize returns how many bytes of memory this response use
+func (rs *batchCopResponse) MemSize() int64 {
+	if rs.respSize != 0 {
+		return rs.respSize
+	}
+
+	// ignore rs.err
+	rs.respSize += int64(cap(rs.startKey))
+	if rs.detail != nil {
+		rs.respSize += int64(sizeofExecDetails)
+		if rs.detail.CommitDetail != nil {
+			rs.respSize += int64(sizeofCommitDetails)
+		}
+	}
+	if rs.pbResp != nil {
+		// Using a approximate size since it's hard to get a accurate value.
+		rs.respSize += int64(rs.pbResp.Size())
+	}
+	return rs.respSize
+}
+
+func (rs *batchCopResponse) RespTime() time.Duration {
+	return rs.respTime
+}
+
+type copTaskAndRPCContext struct {
+	task *copTask
+	ctx  *RPCContext
+}
+
+func getFlashRPCContextWithRetry(bo *Backoffer, task *copTask, cache *RegionCache) (*RPCContext, error) {
+	for {
+		ctx, err := cache.GetTiFlashRPCContext(bo, task.region)
+		if err != nil {
+			return nil, err
+		}
+		if ctx != nil {
+			return ctx, nil
+		}
+
+		if err = bo.Backoff(BoRegionMiss, errors.New("failed to get flash rpc context")); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+}
+
+func buildBatchCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, req *kv.Request) ([]*batchCopTask, error) {
+	if req.StoreType != kv.TiFlash {
+		return nil, errors.New("store type must be tiflash")
+	}
+
+	start := time.Now()
+	const cmdType = tikvrpc.CmdBatchCop
+	rangesLen := ranges.len()
+	var tasks []*copTask
+	appendTask := func(regionWithRangeInfo *KeyLocation, ranges *copRanges) {
+		tasks = append(tasks, &copTask{
+			region:    regionWithRangeInfo.Region,
+			ranges:    ranges,
+			cmdType:   cmdType,
+			storeType: req.StoreType,
+		})
+	}
+
+	err := splitRanges(bo, cache, ranges, appendTask)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var batchTasks []*batchCopTask
+
+	storeTaskMap := make(map[string]*batchCopTask)
+
+	for _, task := range tasks {
+		rpcCtx, err := getFlashRPCContextWithRetry(bo, task, cache)
+		if err != nil {
+			return nil, err
+		}
+		if batchCop, ok := storeTaskMap[rpcCtx.Addr]; ok {
+			batchCop.copTasks = append(batchCop.copTasks, copTaskAndRPCContext{task: task, ctx: rpcCtx})
```

Comment:
Although TiFlash doesn't support descent scan table, we'd reverse the tasks when `req.desc` is true, thus we can obviously find the error when planner pushes the descent order to TiFlash or TiKV using the batch cop request in future.

Reply:
Shall we report error when `KeepOrder` or `Desc` is set to true?